File: 백준/Gold/9663. N－Queen/N－Queen.py
import sys

input = sys.stdin.readline
output = sys.stdout.write

visited1 = [False] * 40
visited2 = [False] * 40
visited3 = [False] * 40

n = int(input())
cnt = 0
def nqueen(cur):
  if cur == n:
    global cnt
    cnt += 1
    return

  for i in range(n):
    if visited1[i] or visited2[i + cur] or visited3[cur-i+n-1] :
      continue
    visited1[i] = True
    visited2[i + cur] = True
    visited3[cur-i+n-1] = True
    nqueen(cur + 1)
    visited1[i] = False
    visited2[i+cur] = False
    visited3[cur-i+n-1] = False

nqueen(0)
output(str(cnt))

File: 프로그래머스/lv2/12952. N－Queen/N－Queen.cpp
#include <string>
#include <vector>

using namespace std;
bool visited1[40] = {false,};
bool visited2[40] = {false,};
bool visited3[40] = {false,};
int cnt = 0;
void nqueen(int cur,int n){
  if (cur == n){
    cnt += 1;
    return;
  }

  for(int i = 0;i<n;i++){
    if (visited1[i] || visited2[i + cur] || visited3[cur-i+n-1])
        continue;
    visited1[i] = true;
    visited2[i + cur] = true;
    visited3[cur-i+n-1] = true;
    nqueen(cur + 1,n);
    visited1[i] = false;
    visited2[i+cur] = false;
    visited3[cur-i+n-1] = false;
  }
}
int solution(int n) {
    int answer = 0;
    nqueen(0,n);
    return cnt;
}